Added a default timeout to Tradfri observations

alex3305 opened the pull request:
## Description:

The IKEA Tradfri devices were configured by default that the observation of the
individual devices never timed out. These observations are used to check the
current state of the device.

However, I have experienced that having an infinite observation, there is a
fair possibility that devices aren't responsive any more from the UI. This
seem to be caused by a race condition somewhere in the async code of Home
Assistant or pytradfri. Or possibly even the underlying apicoap library.

Since setting states through automation still seems to work and after
debugging for a couple of hours, I figured a workaround to this issue was the
least I could contribute. As I was unable to find the root cause. This at least
(partially) solves #9822 and #14386 and is almost equal to the proposal of
@max-te, but with a less frequent observation timeout.

**Related issue (if applicable):** partially fixes #9822 and #14386

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

> Other checks were not applicable.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

alex3305 committed:
Added a default timeout to Tradfri observations

The IKEA Tradfri devices were configured by default that the observation of the
individual devices never timed out. These observations are used to check the
current state of the device.

However, I have experienced that having an infinite observation, there is a
fair possibility that devices aren't responsive any more from the UI. This
seem to be caused by a race condition somewhere in the async code of Home
Assistant or pytradfri. Or possibly even the underlying apicoap library.

Since setting states through automation still seems to work and after
debugging for a couple of hours, I figured a workaround to this issue was the
least I could contribute. As I was unable to find the root cause. This at least
(partially) solves #9822 and #14386 and is almost equal to the proposal of
@max-te, but with a less frequent observation timeout.

homeassistant commented:
Hello @alex3305,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ffde81540ba12c896ba8105973dcd4ad26a3e4a8`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18497&#x2F;commits&#x2F;ffde81540ba12c896ba8105973dcd4ad26a3e4a8) - This commit has something that looks like an email address (alex.van.den.hoogen@geodan.nl). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @alex3305, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18497).

Once you do that we will be able to review and accept this pull request.

Thanks!

alex3305 commented on homeassistant/components/switch/tradfri.py:
```diff
@@ -114,7 +114,7 @@ def _async_start_observe(self, exc=None):
             cmd = self._switch.observe(callback=self._observe_update,
                                        err_callback=self._async_start_observe,
-                                       duration=0)
+                                       duration=DEFAULT_OBSERVE_TIMEOUT)
```
Fixed with amend commit.

alex3305 commented:
Appearently Travis failed because I forgot to commit a variable. I amended that with an amended commit, but was not picked up by Travis unfortunately. Is there anyway to trigger Travis again?

Also it seems that the code that Travis fails on is unrelated.

lwis commented:
It's a shame there are some oddities with the gateway and CoAP, when the connection dies the observation should end and be restarted automatically. While I don't (and have never) experienced this issue on my network, with my gateway, I understand that others do have issues with holding persistent connections to the gateway.

max-te commented:
This differs from my workaround in that you don't call something like ```self.hass.loop.call_later(TIMEOUT, self._async_start_observe)```. I don't see how, in your code, a new observation is started after the timeout is reached.

lwis commented:
@max-te good spot, I'm not sure aiocoap will execute a callback when the observation completes. 

@alex3305 have you tested this?

alex3305 commented:
@lwis I've not tested this extensively yet. I was hoping to see multiple people testing this. The `aiocoap` project didn't have a release in more than a year. Since the changes are quite extensive, the time it takes to trickle to through HASS would be quite long I suppose. This is just a workaround to ensure that I don't have to reboot my HASS install everyday :).

@max-te In `libcoap` when the duration timeout is reached, the `err_callback` function will be called and the timeout will be restarted. I don't know if this is the case with the `aiocoap` library. This can either be amended in Home Assistant or in pytradfri. What do you think?

max-te commented:
At the time I was under the impression that ```err_callback``` isn't called when timeout is reached, I don't remember how I came to that conclusion though, so you might want to test this.

alex3305 committed:
Added an additional `loop_call_later` for Tradfri

Amends #18497 with an additional call to `loop_call_later`. According to
a little more reasearch regarding this PR, me and @max-te saw that the
`err_callback` wasn't called when the set duration timed out.

Maybe this can be fixed in pytradfri, or as @lwis suggested there probably
should be a heartbeat in pytradfri to prevent this kind of behaviour.

Although this workaround works, this change can also cause a bit of a stack
leak with Tradfri devices. But since the timeout is set at 1 hour at default,
this shouldn't be much of an issue.

lwis commented:
Can you make this configurable with the default set to 0?

alex3305 commented:
@lwis I cannot figure out how to get the configurable value working. With all the async threads being passed around and separate classes, it's quite hard to wrap my head around it. 

Can you give me any pointers?

pvizeli commented on homeassistant/components/tradfri/__init__.py:
```diff
@@ -26,6 +26,7 @@
 KEY_API = 'tradfri_api'
 CONF_ALLOW_TRADFRI_GROUPS = 'allow_tradfri_groups'
 DEFAULT_ALLOW_TRADFRI_GROUPS = False
+DEFAULT_OBSERVE_TIMEOUT = 3600  # Set default timeout to 1 hour in seconds
```
The name suggests that is a default for options. Call it `TIMEOUT_OBSERVE`

alex3305 replied:
I'm looking into it, also with @lwis suggestion to make it configurable. 

But thanks for the suggestion.
